src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -288,6 +290,15 @@ const absl::flat_hash_map<NodeID, SchedulingResources>
   return cluster_scheduling_resources_;
 }
 
+std::shared_ptr<rpc::ResourceUsageBatchData> GcsResourceManager::GetAllResourceUsage()
+    const {
+  auto data = std::make_shared<rpc::ResourceUsageBatchData>();
+  for (const auto &pair : node_resource_usages_) {
+    data->add_batch()->CopyFrom(pair.second);
+  }
```

Comment:
I'm a little concerned this will add a lot of CPU overhead to be serializing this message many times for each node.

Maybe it's fine, but if not, we can consider serializing it once, and sending the same binary coded to each node and decoding it explicitly on the raylet side.

Do you have any perf benchmarks on how much overhead this adds?

Reply:
Could also cache this in the GcsResourceManager and overwrite on a change.

Reply:
> Could also cache this in the GcsResourceManager and overwrite on a change.

Hmm I'm a little worried about thread safety with that, since we send and receive resource reports on different threads and they would both share a pointer and access the same memory (if I'm understanding your suggestion correctly). 

There are some optimizations that come to mind.

1. Call the function less (I made a note about this in `TryPullResourceReport`. 
2. Incorporate some concept of last update time to avoid copying the whole batch (this might not help that much)
3. Do the cache idea but with a double buffer, which would introduce some (configurable) staleness. 

I'll run some benchmarks and gather some data before doing anything though.